wxGTK: use GTK 3 by default

lopsided98 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

wxGTK should switch to using GTK 3 where possible. Currently this PR implements the basic change for `wxGTK30` and wxPython 3.0. There are a lot of packages that may be affected by this change, as shown in the following `nix-review` report:
<details>
  <summary>24 packages failed to build:</summary>

  - audacity 
  - boinc 
  - dolphinEmu (refuses to add GTK3 support: https://bugs.dolphin-emu.org/issues/9414)
  - ~~electricsheep~~ (previously broken)
  - filezilla 
  - fityk (Debian uses patch: https://salsa.debian.org/science-team/fityk/blob/master/debian/patches/permit-gtk3.patch)
  - golly-beta (old version, latest is fine; can this be removed?)
  - grass (intermittent failure, built fine when retried, GUI seems to be broken on master)
  - opencpn 
  - pcsx2 (1.4 doesn't officially support GTK3, master does, see: https://github.com/PCSX2/pcsx2/issues/2312)
  - perl528Packages.Wx 
  - perl528Packages.WxGLCanvas 
  - perl530Packages.Wx 
  - perl530Packages.WxGLCanvas 
  - poedit 
  - prusa-slicer 
  - rigsofrods 
  - salut_a_toi 
  - sasview 
  - slade 
  - sladeUnstable 
  - slic3r 
  - wxsqliteplus 
  - xchm 
</details>
<details>
  <summary>89 packages were built:</summary>

  - aegisub 
  - bossa 
  - cl 
  - couchdb 
  - couchdb2 
  - cubicsdr 
  - curaByDagoma 
  - cura_stable 
  - cuter 
  - digikam 
  - displaycal 
  - dvdstyler 
  - ejabberd 
  - elixir 
  - elixir_1_5 
  - elixir_1_6 
  - elixir_1_7 
  - elixir_1_8 
  - erlang 
  - erlangR18 
  - erlangR19 
  - erlangR20 
  - erlangR21 
  - erlang_javac 
  - erlang_odbc 
  - erlang_odbc_javac 
  - far2l 
  - flamerobin 
  - gnss-sdr 
  - gnuradio 
  - gr-ais 
  - gr-gsm 
  - gr-limesdr 
  - gr-nacl 
  - gr-osmosdr 
  - gr-rds 
  - gnuradio-with-packages 
  - golly 
  - gqrx 
  - hex2nix 
  - hugin 
  - inspectrum 
  - kicad 
  - kicad-with-packages3d 
  - lenmus 
  - limesuite 
  - loxodo 
  - lutris 
  - lutris-free 
  - mercury 
  - metamorphose2 
  - mmex 
  - notmuch-bower 
  - odamex 
  - perl528Packages.AlienWxWidgets 
  - perl530Packages.AlienWxWidgets 
  - playonlinux 
  - plover.stable 
  - printrun 
  - python27Packages.pyspread 
  - python27Packages.robotframework-ride 
  - python27Packages.runsnakerun 
  - python27Packages.wxPython 
  - qradiolink 
  - rabbitmq-server 
  - radiotray-ng 
  - rebar 
  - rebar3 
  - relxExe 
  - rtl_433 
  - saga 
  - scyther 
  - sdrangel 
  - soapysdr-with-plugins 
  - torchat 
  - truecrypt 
  - tsung 
  - urh 
  - welle-io 
  - wings 
  - winpdb 
  - winusb 
  - wxGTK30 
  - wxSVG 
  - wxmaxima 
  - wxsqlite3 
  - xylib 
  - yaws 
  - zeroad 
</details>

Some of these failures may be prexisting, but many are caused by the switch to GTK 3.

I would like to drop support for GTK 2 altogether in wxGTK, but I'm not sure if this will be possible. I also need to test wxGTK 3.1 with GTK 3.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lopsided98 committed:
python2Packages.wxPython: use GTK3

lopsided98 committed:
wxGTK: use GTK3 by default

lopsided98 committed:
boinc: support GTK3

lopsided98 committed:
filezilla: support GTK3

Evils-Devils commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5268,7 +5268,7 @@ in {
 
   wxPython30 = callPackage ../development/python-modules/wxPython/3.0.nix {
-    wxGTK = pkgs.wxGTK30;
+    wxGTK = pkgs.wxGTK30.override { withGtk2 = false; };
```
isn't this exactly what you're trying to avoid?
```suggestion
    wxGTK = pkgs.wxGTK30;
```

lopsided98 committed:
opencpn: support GTK3

Evils-Devils commented:
slic3r fixed [here](https://github.com/Evils-Devils/nixpkgs/commit/2244b0e684348b5ae906917bccbda945f0a5d2ab), added hicolor-icon-theme while i was at it.

lopsided98 committed:
perlPackages.AlienWxWidgets: support GTK3

lopsided98 committed:
perlPackages.Wx: support GTK3

lopsided98 committed:
perlPackages.WxGLCanvas: support GTK3

evils committed:
poedit: gtk2 -> gtk3

evils committed:
rigsofrods: wxGTK30 withGtk2 = true

evils committed:
wxGTK31: default to gtk3

lopsided98 committed:
xchm: support GTK3

Evils-Devils commented:
i've addressed some of the failures [here](https://github.com/Evils-Devils/nixpkgs/tree/wxwidgets-gtk3)
i set `withGtk2 = true;` where it seemed appropriate (where the source doesn't mention gtk3),
though maybe using wxGTK29 is more appropriate for those?

i also defaulted wxGTK31 to gtk3, nothing was negatively affected.

i'm still seeing these failures
- dolphinEmu; @delroth seems involved in that, perhaps they can offer clarification?
- prusa-slicer; wxWidgets not found
- slade, sladeUnstable; should support gtk3 but can't find it
- wxsqliteplus; is not fixed by setting `withGtk2 = true;`

and these which are also broken without this PR
- electricsheep
- pcsx2
- salut_a_toi
- sasview

delroth commented:
Re: dolphinEmu, the UI was migrated from WX to Qt a few years ago. Unfortunately I don't think any stable release has been tagged since that migration to Qt. I would either keep wxGTK2 or drop dolphinEmu stable from nixpkgs in favor of dolphinEmuMaster.

lopsided98 committed:
dolphinEmu: use wxGTK with GTK2

lopsided98 commented:
`rigsofrods` has a similar problem; master supports GTK 3, but there has not been an official release since 2016. It seems like upstream considered 4.8rc5 to be a stable release. I tried to update it, but ran into a bunch of runtime issues (crashes, missing textures), so I think I'll just have it use GTK2 for now.

I don't know how aggressive we should be in switching to GTK3. Should I apply the patches Debian and Fedora have written to add GTK3 support to the packages that don't officially support it?

worldofpeace commented:
> I don't know how aggressive we should be in switching to GTK3. Should I apply the patches Debian and Fedora have written to add GTK3 support to the packages that don't officially support it?

@lopsided98 Absolutely, as a gtk maintainer in nixpkgs I'm always glad to come upon these patches where needed :+1:  Though if the package appears completely unused and unmaintained, feel free to remove it in a separate PR.

evils committed:
prusa-slicer: gtk2 -> gtk3
also fix the file chooser

remaining issues
in nix-shell --pure
  fails with a "Switching prusaSlicer to language en_US failed" popup
  and "dbind-WARNING AT-SPI: error retrieving accessibility bus address"

lopsided98 committed:
slade: 3.1.1.5 -> 3.1.7, support GTK3

evils committed:
slade-git: remove, stable is significantly ahead

this seems to have be done in anticipation of the 3.1.2 release
which has come and gone
  stable is at 3.1.7 and master is only 6 commits ahead

evils committed:
wxsqliteplus: use gtk3

build reports as gtk1, source doesn't mention gtk3

wrap to get file chooser

remaining nitpicks from nix-shell --pure
  dbind-WARNING: AT-SPI: Error retrieving accessibility bus address
  dconf-WARNING: failed to commit changes to dconf

c0bw3b commented:
@lopsided98 you can drop audacity commit since it was done in #73055

lopsided98 committed:
prusa-slicer: run tests

lopsided98 committed:
fityk: apply patch to support GTK3

lopsided98 commented:
I think that is everything except for `golly-beta`. The only packages left using wxGTK with GTK 2 are `dolphinEmu` and `rigsofrods`. Both will be fixed next time there is a new upstream release.

In the case of `rigsofrods`, I am experimenting with packaging 0.4.8rc5, which supports GTK 3 and seems to be considered stable by the upstream project. They appear to be planning to switch to a date based versioning scheme and make more frequent releases (https://github.com/RigsOfRods/rigs-of-rods/issues/2415), although I'm not a fan of their choice to automatically set the version to the build date (https://github.com/RigsOfRods/rigs-of-rods/pull/2422). The version currently packaged has some visual bugs (ground textures not rendering) that make many maps almost unplayable, and the problems are even worse with the latest version. These bugs are due to some kind of build issue, but I haven't figured it out yet. If I solve this problem, I will make a separate PR.

I modified some of @Evils-Devils commits to take into account @worldofpeace's comment [here](https://github.com/Evils-Devils/nixpkgs/commit/d4c98c9cc445a9857a030b2a8ce2b6d11efb3665#r35902933). I don't know much about GNOME packaging, and adding different GNOME packages as dependencies does not seem to have much effect on the non-NixOS system I am using for testing, so I would like someone to review them. I did notice that using `wrapGAppsHook` causes the applications to not respect my theme choice.

I would like to remove `golly-beta` if possible, but I don't know there is a reason to keep it. cc @7c6f434c, who added it.

I am also ccing the maintainers of all packages I touched who were not automatically requested for review by ofborg:
prusa-slicer: @thorstenweber83
slade-git: @ertes
rigsofrods: @7c6f434c 
poedit: @bignaux

worldofpeace commented:
> I modified some of @Evils-Devils commits to take into account @worldofpeace's comment here. I don't know much about GNOME packaging, and adding different GNOME packages as dependencies does not seem to have much effect on the non-NixOS system I am using for testing, so I would like someone to review them. I did notice that using wrapGAppsHook causes the applications to not respect my theme choice.

As I just happen to be a Gnome maintainer I'd be happy to do this :+1:

worldofpeace commented on pkgs/development/libraries/wxsqliteplus/default.nix:
```diff
@@ -8,10 +9,12 @@ stdenv.mkDerivation rec {
-  buildInputs = [ wxGTK wxsqlite3 sqlite ];
+  nativeBuildInputs = [ wrapGAppsHook ];
+
+  buildInputs = [ wxGTK wxsqlite3 sqlite gnome3.glib ];
```
```suggestion
  buildInputs = [ wxGTK wxsqlite3 sqlite glib ];
```

old alias

worldofpeace reviewed (changes requested):
You should probably add `wrapGAppsHook` to all the applications.

7c6f434c commented:
OK to drop `golly-beta` now that it is way behind the stable release.
OK whatever happens with `rigsofrods`, it never was packaged in a fully stable way, so if it builds and starts, it's not a big regression.

lopsided98 committed:
golly-beta: remove because stable golly is newer

lopsided98 commented:
Is `wrapGAppsHook` only needed for GTK 3 applications? If we don't apply it to all applications that use wxGTK, will this PR cause regressions for the applications that used to use GTK 2? It seems infeasible to fix all the applications affected by this PR right now, since there are over 100 of them.

Evils-Devils commented:
i switched my branch to lopsided's, rebased on nixos-unstable and tried to execute the rebuild packages via nix-review
<details>
<summary>109 rebuild</summary>
aegisub audacity boinc bossa cl couchdb couchdb2 cubicsdr curaByDagoma cura_stable cuter diff-pdf digikam displaycal dvdstyler ejabberd elixir elixir_1_5 elixir_1_6 elixir_1_7 elixir_1_8 erlang erlangR18 erlangR19 erlangR20 erlangR21 erlang_javac erlang_odbc erlang_odbc_javac far2l filezilla fityk flamerobin gnss-sdr gnuradio gr-ais gr-gsm gr-limesdr gr-nacl gr-osmosdr gr-rds gnuradio-with-packages golly gqrx grass hugin inspectrum kicad kicad-with-packages3d lenmus limesuite loxodo lutris lutris-free mercury metamorphose2 mmex notmuch-bower odamex opencpn perl528Packages.AlienWxWidgets perl528Packages.Wx perl528Packages.WxGLCanvas perl530Packages.AlienWxWidgets perl530Packages.Wx perl530Packages.WxGLCanvas playonlinux plover.stable poedit printrun prusa-slicer pwsafe python27Packages.pyspread python27Packages.robotframework-ride python27Packages.runsnakerun python27Packages.wxPython qradiolink rabbitmq-server radiotray-ng rebar rebar3 relxExe rtl_433 saga scyther sdrangel slade slic3r soapysdr-with-plugins torchat treesheets truecrypt tsung urh welle-io wings winpdb winusb wxGTK30 wxGTK31 wxSVG wxhexeditor wxmaxima wxsqlite3 wxsqliteplus xchm xylib yaws zeroad
</details>
<details>
<summary>28 with file chooser / GSettings-schema crash</summary>
aegisub audacity bossa displaycal dvdstyler filezilla fityk golly hugin kicad lenmus limesuite loxodo metamorphose2 mmex opencpn plover pwsafe python27Packages.runsnakerun saga scyther slic3r treesheets winpdb wxmaxima xchm xylib zeroad
</details>
<details>
<summary>4 with the gtk2 file chooser?</summary>
digikam qradiolink sdrangel urh
</details>
<details>
<summary>5 fail with "no module"</summary>
no module named wx:<br />
curaByDagorna playonlinux torchat<br />
no module named gtk:<br />
gnuradio-companion<br />
no module named _multiarray_umath:<br />
cura
</details>
<details>
<summary>5 seem to work</summary>
far2l lutris poedit slade wxsqliteplus
</details>
<details>
<summary>3 seem to work due to not using/needing a file chooser</summary>
boinc welle-io radiotray-ng
</details>
<details>
<summary>3 seem messed up without this PR as well</summary>
notmuch-bower odamex flamerobin
</details>
<details>
<summary>4 unchanged build failures unrelated to this PR</summary>
electricsheep pcsx2 salut_a_toi sasview
</details>
<details>
<summary>the rest i couldn't test on their own</summary>
couchdb couchdb2 cubicsdr cuter diff-pdf ejabberd elixir* erlang* gnss-sdr gr-* gqrx grass inspectrum mercury perl5* printrun prusa-slicer python27* rabbitmq-server rebar rebar3 relxExe rtl_433 soapysdr-with-plugins truecrypt tsung wings winusb wxGTK3* wxSVG wxhexeditor wxsqlite3 yaws
</details>

@worldofpeace since most of the packages using wxGTK30 with gtk3 seem to be affected, is there a more systemic way to treat this rather than to patch each of them?

worldofpeace commented:
when you get
`No GSettings schemas are installed on the system`

see https://nixos.org/nixpkgs/manual/#ssec-gnome-common-issues and the rest of the GNOME docs.

It needs `wrapGAppsHook` in `nativeBuildInputs` and some other package that provides the gsettings schema. I'm guessing it's probably the filechooser so you'll probably want `wxGTK.gtk`.

@lopsided98 The gtk2 version of the apps maybe used GConf, so yes, doing I've instructed in the previous comment is the only way to land this PR without regressions.

It also seems @Evils-Devils has done a great job of testing it as well.

worldofpeace reviewed (changes requested):
See above comment.

worldofpeace commented:
>  is there a more systemic way to treat this rather than to patch each of them

No, that's simply how packaging gtk3 apps works in nixpkgs. It all revolves around that setup hook.

peterhoeg commented:
https://github.com/NixOS/nixpkgs/pull/86217 moves filezilla to use gtk3 (and thus wxgtk with gtk3 by extension).

peterhoeg commented:
wxwidgets 3.0.5 is out as of y'day which also updates gstreamer from 0.10 to 1.0 which would be great move.

jonringer commented:
@lopsided98 I know that this PR is probably very tiring as a lot is going on the master branch. If you want to break this apart into smaller PRs, it will also make them easier to merge as there's less surface area to test

lopsided98 commented:
Yeah, I think that is the only feasible way to move forward with this. We can just make individual packages use GTK3, and then someday go back and change the default.

I don't have a lot of time to work on this right now, but if others want to take commits from this PR and make separate PRs, feel free to go ahead.

peterhoeg commented:
How about adding -gtk2 and -gtk3 variants to all-packages.nix and have the main wxgtk just be an alias to -gtk2 for now. This makes the transition very clear.
